src/idpi/grib_decoder.py

```diff
@@ -82,6 +52,171 @@ def _extract_pv(pv):
+            }
+
+    def _gather_coords(self):
+        if self.dims is None:
+            raise RuntimeError("No dims.")
+
+        coord_values = zip(*self.values)
+        unique = (sorted(set(values)) for values in coord_values)
+        coords = {dim: c for dim, c in zip(self.dims[:-2], unique)}
+
+        if missing := [
+            combination
+            for combination in product(*coords.values())
+            if combination not in self.values
+        ]:
+            msg = f"Missing combinations: {missing}"
+            logger.exception(msg)
+            raise RuntimeError(msg)
+
+        ny, nx = next(iter(self.values.values())).shape
+        shape = tuple(len(v) for v in coords.values()) + (ny, nx)
+        return coords, shape
+
+    def _gather_tcoords(self):
+        time = None
+        valid_time = []
+        for step in sorted(self.time_meta):
+            tm = self.time_meta[step]
+            valid_time.append(_parse_datetime(tm["validityDate"], tm["validityTime"]))
+            if time is None:
+                time = _parse_datetime(tm["dataDate"], tm["dataTime"])
+
+        return {"valid_time": ("time", valid_time), "ref_time": time}
+
+    def to_xarray(self) -> xr.DataArray:
+        if not self.values:
+            raise RuntimeError("No values.")
+
+        coords, shape = self._gather_coords()
+        tcoords = self._gather_tcoords()
+
+        array = xr.DataArray(
+            data=np.array(
+                [self.values.pop(key) for key in sorted(self.values)]
+            ).reshape(shape),
+            coords=coords | self.hcoords | tcoords,
+            dims=self.dims,
+            attrs=self.metadata,
+        )
+
+        if array.vcoord_type != "surface":
+            return array
+
+        return array.squeeze("z", drop=True)
+
+
+def _load_buffer_map(
+    source: data_source.DataSource,
+    request: Request,
+) -> dict[str, _FieldBuffer]:
+    logger.info("Retrieving request: %s", request)
+    fs = source.retrieve(request)
+
+    buffer_map: dict[str, _FieldBuffer] = {}
+
+    for field in fs:
+        name = field.metadata("shortName")
+        buffer = buffer_map.setdefault(name, _FieldBuffer())
+        buffer.load(field)
+
+    return buffer_map
+
+
+def load_single_param(
+    source: data_source.DataSource,
+    request: Request,
+) -> xr.DataArray:
+    """Request data from a data source for a single parameter.
+
+    Parameters
+    ----------
+    source : data_source.DataSource
+        Source to request the data from.
+    request : str | tuple[str, str] | dict[str, Any]
+        Request for data from the source in the mars language.
+
+    Raises
+    ------
+    RuntimeError
+        when data is missing.
+
+    Returns
+    -------
+    xarray.DataArray
+        A data array of the requested field.
+
+    """
+    buffer_map = _load_buffer_map(source, request)
+    [buffer] = buffer_map.values()
+    return buffer.to_xarray()
+
+
+def load(
```

Comment:
This function is the new API that allows to pass multiple params in a single request so it is intended for the users of the library. It should indeed have a unit test. The `GribReader.load` function now uses the `load_single_param` but does not replace it. It is set up this way to avoid introducing a breaking change now and we might need to consider deprecating the GribReader class in the future.